panther_hardware_interfaces/src/panther_imu.cpp

```diff
@@ -0,0 +1,513 @@
+}
+
+void PantherImuSensor::Calibrate()
+{
+  if (spatial_ == nullptr) {
+    throw std::runtime_error("Imu hardware is not active yet!");
+  }
+
+  spatial_->zero();
+  // The API call returns directly, so we "enforce" the recommended 2 sec
+  // here. See: https://github.com/ros-drivers/phidgets_drivers/issues/40
+
+  // FIXME: Ideally we'd use an rclcpp method that honors use_sim_time here,
+  // but that doesn't actually exist.  Once
+  // https://github.com/ros2/rclcpp/issues/465 is solved, we can revisit this.
+  std::this_thread::sleep_for(std::chrono::seconds(2));
+}
+
+bool PantherImuSensor::IsParamDefined(const std::string & param_name)
+{
+  return info_.hardware_parameters.find(param_name) != info_.hardware_parameters.end();
+}
+
+void PantherImuSensor::ConfigureCompassParams()
+{
+  if (
+    IsParamDefined("cc_mag_field") and IsParamDefined("cc_offset0") and
+    IsParamDefined("cc_offset1") and IsParamDefined("cc_offset2") and IsParamDefined("cc_gain0") and
+    IsParamDefined("cc_gain1") and IsParamDefined("cc_gain2") and IsParamDefined("cc_t0") and
+    IsParamDefined("cc_t1") and IsParamDefined("cc_t2") and IsParamDefined("cc_t3") and
+    IsParamDefined("cc_t4") and IsParamDefined("cc_t5")) {
+    ReadCompassParams();
+
+    spatial_->setCompassCorrectionParameters(
+      params_.cc_mag_field, params_.cc_offset0, params_.cc_offset1, params_.cc_offset2,
+      params_.cc_gain0, params_.cc_gain1, params_.cc_gain2, params_.cc_t0, params_.cc_t1,
+      params_.cc_t2, params_.cc_t3, params_.cc_t4, params_.cc_t5);
+  } else {
+    RCLCPP_INFO(logger_, "No compass correction params found. Skipping...");
+  }
+}
+
+void PantherImuSensor::ConfigureHeating()
+{
+  if (IsParamDefined("heating_enabled")) {
+    params_.heating_enabled =
+      hardware_interface::stod(info_.hardware_parameters["heating_enabled"]);
+    spatial_->setHeatingEnabled(params_.heating_enabled);
+  } else {
+    RCLCPP_INFO(logger_, "No heating enabled found. Skipping...");
+  }
+}
+
+void PantherImuSensor::ConfigureMadgwickFilter()
+{
+  ReadMadgwickFilterParams();
+
+  filter_.setWorldFrame(world_frame_);
+  filter_.setAlgorithmGain(params_.gain);
+  filter_.setDriftBiasGain(params_.zeta);
+}
+
+void PantherImuSensor::SpatialDataCallback(
+  const double acceleration[3], const double angular_rate[3], const double magnetic_field[3],
+  double timestamp)
+{
+  auto timestamp_s = timestamp * 1e-3;
+  RCLCPP_DEBUG(logger_, "SpatialDataCallback...");
+
+  std::lock_guard<std::mutex> lock(spatial_mutex_);
+
+  geometry_msgs::msg::Vector3 mag_fld;
+  geometry_msgs::msg::Vector3 ang_vel;
+  geometry_msgs::msg::Vector3 lin_acc;
+
+  mag_fld.x = magnetic_field[0] * 1e-4;
+  mag_fld.y = magnetic_field[1] * 1e-4;
+  mag_fld.z = magnetic_field[2] * 1e-4;
+
+  for (auto i = 0u; i < 3; ++i) {
+    if (magnetic_field[i] == KImuMagneticFieldUnknownValue) {
+      mag_fld.x = std::numeric_limits<double>::quiet_NaN();
+    }
+  }
```

Comment:
What is the purpose of setting only the x value to NaN? Why do we perform this operation, and then immediately check if x is finite in the following lines?

Reply:
Ohhhh.. It is a mistake! All values should be nan. Thank you for that catch.